# Network construction : from gene expression data to co-expression network (visualization).
library(WGCNA)
library(igraph)
library(EDASeq)
library(pheatmap)
library(DESeq2)

# Step1: Read in gene expression data.
# Load sample information "sample_info.used" and gene expression profiles "txi.salmon.full.used".
# Gene expression profile (txi.salmon.full.used) is generated by salmon and read in with DESeq2. You can use your favorite \n
#   RNAseq quantification software.
load("ant_queen_exp.Rdata") # Expression matrix
load("ant_queen_DEGs.Rdata") # DEG (deferentially expressed genes) data

# Step2: Normalize expression data and filter data.
# Perform quantile normalization and log transformation on the expression profiles, so that the\n
# expression profiles are comparable across samples and the gene-gene correlation \n
# will not be driven by few highly expressed samples (it will matter if we use Pearson correlation \n,
# or Euclean distance to measure gene-gene distance, but won't matter when use Spearman's correlation.
exp.quantile = betweenLaneNormalization(txi.salmon.full.used$abundance,
                                        offset=FALSE, round=F,which = 'full')
exp.quantile = log2(exp.quantile +1)

# Data filtering:
# Select target genes to construct co-expression network, using genes \n
# that being deferentially expressed for at least one age group.
insemination_gene = rownames(result.mpha.15d)[which(result.mpha.5d$padj < 5e-2 | 
                                                      result.mpha.15d$padj < 5e-2 |
                                                      result.mpha.30d$padj < 5e-2)]
exp.quantile.select = exp.quantile[insemination_gene,]

# Filter genes with 0 variation, as they can't be used to construct co-expression network.
# First calculate the expression variation in each group.
exp_var = data.frame(t(apply(exp.quantile.select,1,function(x) 
  tapply(x,INDEX = paste(sample_info.used$Age,sample_info.used$Treatment),var))))
# Then filter genes without expression variation.
gene_list = rownames(exp_var)[apply(exp_var,1,min) > 0]
exp.quantile.select = exp.quantile.select[gene_list,]

# Step3: Construct co-expression matrix (network) from individual data.

gene.cor_matrix_list = tapply(colnames(exp.quantile.select),
                              INDEX = paste(sample_info.used$Age,sample_info.used$Treatment),
                              function(x){
                                # Calculate gene-gene correlations and 
                                # use absolute value (abs) to include both negative and positive correlation
                                gene.cor_matrix = abs(cor(t(exp.quantile.select[,x]),method = 's'))
                                # Calculate the pvalues for correlations and filter non-significant gene-gene correlation.
                                gene.cor_matrix.p = corPvalueStudent(gene.cor_matrix, nSamples = length(x))
                                gene.cor_matrix.filter = gene.cor_matrix
                                gene.cor_matrix.filter[which(gene.cor_matrix.p > 1e-3)] = 0
                                return(gene.cor_matrix.filter)})

# Step4: Identify key genes in co-expression matrix (network).
# An example: Finding hub genes [genes with highest co-expression correlation with other genes].
# Using day 5 queen as an example.
hubness_day_5 = apply(gene.cor_matrix_list[[paste('5','Queen')]],1, mean,na.rm = T)
hub_genes_day_5 = names(sort(hubness_day_5,decreasing = T))[c(1:30)]

# Visualize the co-expression matrix with 200 genes (randomly select).
gene_annotation = data.frame(row.names = rownames(exp.quantile.select))
gene_annotation$state = 'Background'
gene_annotation[hub_genes_day_5,'state'] = 'Hub genes'
sampled_genes = sample(rownames(exp.quantile.select),200)
pheatmap(gene.cor_matrix_list[[paste('5','Queen')]][sampled_genes,sampled_genes],show_colnames = F,show_rownames = F,
         annotation_row = gene_annotation)

# Step5: Visualize the network.
# Using day 5 queen as an example.
# Subset gene network.
test_age= '5'
test_treatment = 'Queen'
subset_net = paste(test_age,test_treatment)

# Generate gene network from co-expression matrix.
gene_net = graph_from_adjacency_matrix(adjmatrix = gene.cor_matrix_list[[subset_net]],
                                       weighted = T,mode = 'undirected',diag = F)
V(gene_net)$label <- NA # Remove the labels to make it clear.
# Network visualization layout, using force-directed graph drawing, so that genes \n
# with high correlations are clustered together.
l_fr = layout_with_fr(gene_net)

V(gene_net)$frame_color = NA # Setting the dot frame color to transparent.

par(bg="white")
# The links (connections between genes) are visualized with weigths,\n
# so that gene-pairs with high correlation coefficient are with thicker link.
E(gene_net)$width <- E(gene_net)$weight^3/5 
E(gene_net)$color = 'black'
vertex.size.plot = 2
V(gene_net)$color = 'grey'
candidate_list = hub_genes_day_5
gene_color = c('red','grey')
names(gene_color) = c("Hub genes in Day5 queens",'Other genes')
V(gene_net)$color[match(candidate_list,names(V(gene_net)))] = 'red'

# Plotting.
plot(gene_net,vertex.size = vertex.size.plot,
     layout = l_fr,vertex.frame.color = V(gene_net)$frame_color,
     main = paste0('Co-expression network for ',test_treatment,' at day ',test_age))
legend(x=-1.1, y=1.1, legend = names(gene_color) , pch=20,col=gene_color,  pt.cex=1, cex=.5, bty="n", ncol=1)



